File: src/types/shortcut.ts
/**
 * 快捷键平台类型
 */
export type ShortcutPlatform = 'windows' | 'macos' | 'linux';

/**
 * 快捷键按键类型
 */
export interface ShortcutKey {
  /** 按键名称，如 'Ctrl', 'Shift', 'C' */
  key: string;
  /** 按键代码 */
  keyCode: string;
}

/**
 * 快捷键配置项
 */
export interface ShortcutItem {
  /** 唯一标识 */
  id: string;
  /** 显示名称 */
  label: string;
  /** 图标（emoji 或图标名称） */
  icon: string;
  /** 按键组合 */
  keys: ShortcutKey[];
  /** 适用平台 */
  platform: ShortcutPlatform;
  /** 是否启用 */
  enabled: boolean;
  /** 显示顺序 */
  order: number;
}

/**
 * 快捷键设置
 */
export interface ShortcutSettings {
  /** 当前选择的平台 */
  currentPlatform: ShortcutPlatform;
  /** 快捷键列表 */
  shortcuts: ShortcutItem[];
}

/**
 * 快捷键事件
 */
export interface ShortcutEvent {
  /** 快捷键ID */
  shortcutId: string;
  /** 按键组合 */
  keys: ShortcutKey[];
  /** 时间戳 */
  timestamp: number;
}

/**
 * 预设快捷键模板
 */
export const PRESET_SHORTCUTS: Record<ShortcutPlatform, ShortcutItem[]> = {
  windows: [
    {
      id: 'copy',
      label: '复制',
      icon: '📋',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'C', keyCode: 'KeyC' }],
      platform: 'windows',
      enabled: true,
      order: 1,
    },
    {
      id: 'paste',
      label: '粘贴',
      icon: '📄',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'V', keyCode: 'KeyV' }],
      platform: 'windows',
      enabled: true,
      order: 2,
    },
    {
      id: 'save',
      label: '保存',
      icon: '💾',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'S', keyCode: 'KeyS' }],
      platform: 'windows',
      enabled: true,
      order: 3,
    },
    {
      id: 'find',
      label: '查找',
      icon: '🔍',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'F', keyCode: 'KeyF' }],
      platform: 'windows',
      enabled: true,
      order: 4,
    },
    {
      id: 'undo',
      label: '撤销',
      icon: '↶',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'Z', keyCode: 'KeyZ' }],
      platform: 'windows',
      enabled: true,
      order: 5,
    },
    {
      id: 'alt-tab',
      label: '切换窗口',
      icon: '🗔',
      keys: [{ key: 'Alt', keyCode: 'AltLeft' }, { key: 'Tab', keyCode: 'Tab' }],
      platform: 'windows',
      enabled: true,
      order: 6,
    },
    {
      id: 'lock',
      label: '锁屏',
      icon: '🔒',
      keys: [{ key: 'Win', keyCode: 'MetaLeft' }, { key: 'L', keyCode: 'KeyL' }],
      platform: 'windows',
      enabled: true,
      order: 7,
    },
    {
      id: 'task-manager',
      label: '任务管理器',
      icon: '⚡',
      keys: [
        { key: 'Ctrl', keyCode: 'ControlLeft' },
        { key: 'Shift', keyCode: 'ShiftLeft' },
        { key: 'Esc', keyCode: 'Escape' },
      ],
      platform: 'windows',
      enabled: true,
      order: 8,
    },
  ],
  macos: [
    {
      id: 'copy',
      label: '复制',
      icon: '📋',
      keys: [{ key: 'Cmd', keyCode: 'MetaLeft' }, { key: 'C', keyCode: 'KeyC' }],
      platform: 'macos',
      enabled: true,
      order: 1,
    },
    {
      id: 'paste',
      label: '粘贴',
      icon: '📄',
      keys: [{ key: 'Cmd', keyCode: 'MetaLeft' }, { key: 'V', keyCode: 'KeyV' }],
      platform: 'macos',
      enabled: true,
      order: 2,
    },
    {
      id: 'save',
      label: '保存',
      icon: '💾',
      keys: [{ key: 'Cmd', keyCode: 'MetaLeft' }, { key: 'S', keyCode: 'KeyS' }],
      platform: 'macos',
      enabled: true,
      order: 3,
    },
    {
      id: 'find',
      label: '查找',
      icon: '🔍',
      keys: [{ key: 'Cmd', keyCode: 'MetaLeft' }, { key: 'F', keyCode: 'KeyF' }],
      platform: 'macos',
      enabled: true,
      order: 4,
    },
    {
      id: 'undo',
      label: '撤销',
      icon: '↶',
      keys: [{ key: 'Cmd', keyCode: 'MetaLeft' }, { key: 'Z', keyCode: 'KeyZ' }],
      platform: 'macos',
      enabled: true,
      order: 5,
    },
    {
      id: 'cmd-tab',
      label: '切换窗口',
      icon: '🗔',
      keys: [{ key: 'Cmd', keyCode: 'MetaLeft' }, { key: 'Tab', keyCode: 'Tab' }],
      platform: 'macos',
      enabled: true,
      order: 6,
    },
    {
      id: 'lock',
      label: '锁屏',
      icon: '🔒',
      keys: [
        { key: 'Ctrl', keyCode: 'ControlLeft' },
        { key: 'Cmd', keyCode: 'MetaLeft' },
        { key: 'Q', keyCode: 'KeyQ' },
      ],
      platform: 'macos',
      enabled: true,
      order: 7,
    },
    {
      id: 'screenshot',
      label: '截图',
      icon: '📷',
      keys: [
        { key: 'Cmd', keyCode: 'MetaLeft' },
        { key: 'Shift', keyCode: 'ShiftLeft' },
        { key: '4', keyCode: 'Digit4' },
      ],
      platform: 'macos',
      enabled: true,
      order: 8,
    },
  ],
  linux: [
    {
      id: 'copy',
      label: '复制',
      icon: '📋',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'C', keyCode: 'KeyC' }],
      platform: 'linux',
      enabled: true,
      order: 1,
    },
    {
      id: 'paste',
      label: '粘贴',
      icon: '📄',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'V', keyCode: 'KeyV' }],
      platform: 'linux',
      enabled: true,
      order: 2,
    },
    {
      id: 'save',
      label: '保存',
      icon: '💾',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'S', keyCode: 'KeyS' }],
      platform: 'linux',
      enabled: true,
      order: 3,
    },
    {
      id: 'find',
      label: '查找',
      icon: '🔍',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'F', keyCode: 'KeyF' }],
      platform: 'linux',
      enabled: true,
      order: 4,
    },
    {
      id: 'undo',
      label: '撤销',
      icon: '↶',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'Z', keyCode: 'KeyZ' }],
      platform: 'linux',
      enabled: true,
      order: 5,
    },
    {
      id: 'alt-tab',
      label: '切换窗口',
      icon: '🗔',
      keys: [{ key: 'Alt', keyCode: 'AltLeft' }, { key: 'Tab', keyCode: 'Tab' }],
      platform: 'linux',
      enabled: true,
      order: 6,
    },
    {
      id: 'lock',
      label: '锁屏',
      icon: '🔒',
      keys: [{ key: 'Super', keyCode: 'MetaLeft' }, { key: 'L', keyCode: 'KeyL' }],
      platform: 'linux',
      enabled: true,
      order: 7,
    },
    {
      id: 'terminal',
      label: '终端',
      icon: '💻',
      keys: [{ key: 'Ctrl', keyCode: 'ControlLeft' }, { key: 'Alt', keyCode: 'AltLeft' }, { key: 'T', keyCode: 'KeyT' }],
      platform: 'linux',
      enabled: true,
      order: 8,
    },
  ],
};

/**
 * 获取平台的显示名称
 */
export function getPlatformDisplayName(platform: ShortcutPlatform): string {
  const names: Record<ShortcutPlatform, string> = {
    windows: 'Windows',
    macos: 'macOS',
    linux: 'Linux',
  };
  return names[platform];
}

/**
 * 格式化快捷键显示文本
 */
export function formatShortcutKeys(keys: ShortcutKey[]): string {
  return keys.map((k) => k.key).join(' + ');
}

File: lib/models/shortcut.dart
/// 快捷键平台类型
enum ShortcutPlatform {
  windows,
  macos,
  linux,
}

/// 快捷键按键
class ShortcutKey {
  final String key;
  final String keyCode;

  ShortcutKey({required this.key, required this.keyCode});

  Map<String, dynamic> toJson() => {'key': key, 'keyCode': keyCode};

  factory ShortcutKey.fromJson(Map<String, dynamic> json) => ShortcutKey(
        key: json['key'] as String,
        keyCode: json['keyCode'] as String,
      );
}

/// 快捷键配置项
class ShortcutItem {
  final String id;
  final String label;
  final String icon;
  final List<ShortcutKey> keys;
  final ShortcutPlatform platform;
  final bool enabled;
  final int order;

  ShortcutItem({
    required this.id,
    required this.label,
    required this.icon,
    required this.keys,
    required this.platform,
    this.enabled = true,
    required this.order,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'label': label,
        'icon': icon,
        'keys': keys.map((k) => k.toJson()).toList(),
        'platform': platform.name,
        'enabled': enabled,
        'order': order,
      };

  factory ShortcutItem.fromJson(Map<String, dynamic> json) => ShortcutItem(
        id: json['id'] as String,
        label: json['label'] as String,
        icon: json['icon'] as String,
        keys: (json['keys'] as List)
            .map((k) => ShortcutKey.fromJson(k as Map<String, dynamic>))
            .toList(),
        platform: ShortcutPlatform.values.firstWhere(
          (p) => p.name == json['platform'] as String,
          orElse: () => ShortcutPlatform.windows,
        ),
        enabled: json['enabled'] as bool? ?? true,
        order: json['order'] as int? ?? 0,
      );

  ShortcutItem copyWith({
    String? id,
    String? label,
    String? icon,
    List<ShortcutKey>? keys,
    ShortcutPlatform? platform,
    bool? enabled,
    int? order,
  }) =>
      ShortcutItem(
        id: id ?? this.id,
        label: label ?? this.label,
        icon: icon ?? this.icon,
        keys: keys ?? this.keys,
        platform: platform ?? this.platform,
        enabled: enabled ?? this.enabled,
        order: order ?? this.order,
      );
}

/// 快捷键设置
class ShortcutSettings {
  final ShortcutPlatform currentPlatform;
  final List<ShortcutItem> shortcuts;

  ShortcutSettings({
    this.currentPlatform = ShortcutPlatform.windows,
    List<ShortcutItem>? shortcuts,
  }) : shortcuts = shortcuts ?? getDefaultShortcuts(currentPlatform);

  Map<String, dynamic> toJson() => {
        'currentPlatform': currentPlatform.name,
        'shortcuts': shortcuts.map((s) => s.toJson()).toList(),
      };

  factory ShortcutSettings.fromJson(Map<String, dynamic> json) {
    final platformName = json['currentPlatform'] as String? ?? 'windows';
    final platform = ShortcutPlatform.values.firstWhere(
      (p) => p.name == platformName,
      orElse: () => ShortcutPlatform.windows,
    );
    return ShortcutSettings(
      currentPlatform: platform,
      shortcuts: (json['shortcuts'] as List?)
              ?.map((s) => ShortcutItem.fromJson(s as Map<String, dynamic>))
              .toList() ??
          getDefaultShortcuts(platform),
    );
  }

  ShortcutSettings copyWith({
    ShortcutPlatform? currentPlatform,
    List<ShortcutItem>? shortcuts,
  }) =>
      ShortcutSettings(
        currentPlatform: currentPlatform ?? this.currentPlatform,
        shortcuts: shortcuts ?? this.shortcuts,
      );

  /// 获取当前平台的快捷键列表（按顺序排序，仅启用的）
  List<ShortcutItem> get enabledShortcuts =>
      shortcuts.where((s) => s.enabled).toList()
        ..sort((a, b) => a.order.compareTo(b.order));
}

/// 获取平台的显示名称
String getPlatformDisplayName(ShortcutPlatform platform) {
  switch (platform) {
    case ShortcutPlatform.windows:
      return 'Windows';
    case ShortcutPlatform.macos:
      return 'macOS';
    case ShortcutPlatform.linux:
      return 'Linux';
  }
}

/// 格式化快捷键显示文本
String formatShortcutKeys(List<ShortcutKey> keys) {
  return keys.map((k) => k.key).join(' + ');
}

/// 获取默认快捷键列表
List<ShortcutItem> getDefaultShortcuts(ShortcutPlatform platform) {
  switch (platform) {
    case ShortcutPlatform.windows:
      return [
        ShortcutItem(
          id: 'copy',
          label: '复制',
          icon: '📋',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'C', keyCode: 'KeyC')
          ],
          platform: platform,
          order: 1,
        ),
        ShortcutItem(
          id: 'paste',
          label: '粘贴',
          icon: '📄',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'V', keyCode: 'KeyV')
          ],
          platform: platform,
          order: 2,
        ),
        ShortcutItem(
          id: 'save',
          label: '保存',
          icon: '💾',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'S', keyCode: 'KeyS')
          ],
          platform: platform,
          order: 3,
        ),
        ShortcutItem(
          id: 'find',
          label: '查找',
          icon: '🔍',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'F', keyCode: 'KeyF')
          ],
          platform: platform,
          order: 4,
        ),
        ShortcutItem(
          id: 'undo',
          label: '撤销',
          icon: '↶',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'Z', keyCode: 'KeyZ')
          ],
          platform: platform,
          order: 5,
        ),
        ShortcutItem(
          id: 'alt-tab',
          label: '切换窗口',
          icon: '🗔',
          keys: [
            ShortcutKey(key: 'Alt', keyCode: 'AltLeft'),
            ShortcutKey(key: 'Tab', keyCode: 'Tab')
          ],
          platform: platform,
          order: 6,
        ),
        ShortcutItem(
          id: 'lock',
          label: '锁屏',
          icon: '🔒',
          keys: [
            ShortcutKey(key: 'Win', keyCode: 'MetaLeft'),
            ShortcutKey(key: 'L', keyCode: 'KeyL')
          ],
          platform: platform,
          order: 7,
        ),
        ShortcutItem(
          id: 'task-manager',
          label: '任务管理器',
          icon: '⚡',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'Shift', keyCode: 'ShiftLeft'),
            ShortcutKey(key: 'Esc', keyCode: 'Escape'),
          ],
          platform: platform,
          order: 8,
        ),
      ];

    case ShortcutPlatform.macos:
      return [
        ShortcutItem(
          id: 'copy',
          label: '复制',
          icon: '📋',
          keys: [
            ShortcutKey(key: 'Cmd', keyCode: 'MetaLeft'),
            ShortcutKey(key: 'C', keyCode: 'KeyC')
          ],
          platform: platform,
          order: 1,
        ),
        ShortcutItem(
          id: 'paste',
          label: '粘贴',
          icon: '📄',
          keys: [
            ShortcutKey(key: 'Cmd', keyCode: 'MetaLeft'),
            ShortcutKey(key: 'V', keyCode: 'KeyV')
          ],
          platform: platform,
          order: 2,
        ),
        ShortcutItem(
          id: 'save',
          label: '保存',
          icon: '💾',
          keys: [
            ShortcutKey(key: 'Cmd', keyCode: 'MetaLeft'),
            ShortcutKey(key: 'S', keyCode: 'KeyS')
          ],
          platform: platform,
          order: 3,
        ),
        ShortcutItem(
          id: 'find',
          label: '查找',
          icon: '🔍',
          keys: [
            ShortcutKey(key: 'Cmd', keyCode: 'MetaLeft'),
            ShortcutKey(key: 'F', keyCode: 'KeyF')
          ],
          platform: platform,
          order: 4,
        ),
        ShortcutItem(
          id: 'undo',
          label: '撤销',
          icon: '↶',
          keys: [
            ShortcutKey(key: 'Cmd', keyCode: 'MetaLeft'),
            ShortcutKey(key: 'Z', keyCode: 'KeyZ')
          ],
          platform: platform,
          order: 5,
        ),
        ShortcutItem(
          id: 'cmd-tab',
          label: '切换窗口',
          icon: '🗔',
          keys: [
            ShortcutKey(key: 'Cmd', keyCode: 'MetaLeft'),
            ShortcutKey(key: 'Tab', keyCode: 'Tab')
          ],
          platform: platform,
          order: 6,
        ),
        ShortcutItem(
          id: 'lock',
          label: '锁屏',
          icon: '🔒',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'Cmd', keyCode: 'MetaLeft'),
            ShortcutKey(key: 'Q', keyCode: 'KeyQ'),
          ],
          platform: platform,
          order: 7,
        ),
        ShortcutItem(
          id: 'screenshot',
          label: '截图',
          icon: '📷',
          keys: [
            ShortcutKey(key: 'Cmd', keyCode: 'MetaLeft'),
            ShortcutKey(key: 'Shift', keyCode: 'ShiftLeft'),
            ShortcutKey(key: '4', keyCode: 'Digit4'),
          ],
          platform: platform,
          order: 8,
        ),
      ];

    case ShortcutPlatform.linux:
      return [
        ShortcutItem(
          id: 'copy',
          label: '复制',
          icon: '📋',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'C', keyCode: 'KeyC')
          ],
          platform: platform,
          order: 1,
        ),
        ShortcutItem(
          id: 'paste',
          label: '粘贴',
          icon: '📄',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'V', keyCode: 'KeyV')
          ],
          platform: platform,
          order: 2,
        ),
        ShortcutItem(
          id: 'save',
          label: '保存',
          icon: '💾',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'S', keyCode: 'KeyS')
          ],
          platform: platform,
          order: 3,
        ),
        ShortcutItem(
          id: 'find',
          label: '查找',
          icon: '🔍',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'F', keyCode: 'KeyF')
          ],
          platform: platform,
          order: 4,
        ),
        ShortcutItem(
          id: 'undo',
          label: '撤销',
          icon: '↶',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'Z', keyCode: 'KeyZ')
          ],
          platform: platform,
          order: 5,
        ),
        ShortcutItem(
          id: 'alt-tab',
          label: '切换窗口',
          icon: '🗔',
          keys: [
            ShortcutKey(key: 'Alt', keyCode: 'AltLeft'),
            ShortcutKey(key: 'Tab', keyCode: 'Tab')
          ],
          platform: platform,
          order: 6,
        ),
        ShortcutItem(
          id: 'lock',
          label: '锁屏',
          icon: '🔒',
          keys: [
            ShortcutKey(key: 'Super', keyCode: 'MetaLeft'),
            ShortcutKey(key: 'L', keyCode: 'KeyL')
          ],
          platform: platform,
          order: 7,
        ),
        ShortcutItem(
          id: 'terminal',
          label: '终端',
          icon: '💻',
          keys: [
            ShortcutKey(key: 'Ctrl', keyCode: 'ControlLeft'),
            ShortcutKey(key: 'Alt', keyCode: 'AltLeft'),
            ShortcutKey(key: 'T', keyCode: 'KeyT'),
          ],
          platform: platform,
          order: 8,
        ),
      ];
  }
}
